lib/pbench/test/unit/server/test_datasets_metadata.py

```diff
@@ -0,0 +1,178 @@
+                if tmp_req_keys != required_keys:
+                    all_combinations.append(item)
+
+        for items in all_combinations:
+            keys = {}
+            for key, parameter in items:
+                if parameter.type is ParamType.ACCESS:
+                    keys[key] = "public"
+                elif parameter.type is ParamType.DATE:
+                    keys[key] = "2020"
+                elif parameter.type is ParamType.LIST:
+                    keys[key] = []
+                elif parameter.type is ParamType.KEYWORD:
+                    keys[key] = parameter.keywords[0]
+                else:
+                    keys[key] = "foobar"
+
+            missing_key_helper(keys)
+
+        # Test in case all of the required keys are missing and some
+        # random non-existent key is present in the payload
+        if required_keys:
+            missing_key_helper({"notakey": None})
+
+    def test_put_no_dataset(self, client, server_config, attach_dataset):
+        response = client.put(
+            f"{server_config.rest_uri}/datasets/metadata",
+            json={
+                "controller": "node",
+                "name": "foobar",
+                "metadata": {"seen": True, "saved": False},
+            },
+        )
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        assert response.json == {"message": "Dataset foobar not found"}
+
+    def test_put_bad_keys(self, client, server_config, attach_dataset):
+        response = client.put(
+            f"{server_config.rest_uri}/datasets/metadata",
+            json={
+                "controller": "node",
+                "name": "drb",
+                "metadata": {"xyzzy": "private", "what": "sup", "saved": True},
+            },
+        )
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        assert response.json == {
+            "message": "Unrecognized keywords what,xyzzy given for parameter metadata; allowed keywords are saved,seen,user"
+        }
+
+    def test_put_reserved_metadata(self, client, server_config, attach_dataset):
+        response = client.put(
+            f"{server_config.rest_uri}/datasets/metadata",
+            json={
+                "controller": "node",
+                "name": "drb",
+                "metadata": {"access": "private"},
+            },
+        )
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        assert response.json == {
+            "message": "Unrecognized keywords access given for parameter metadata; allowed keywords are saved,seen,user"
+        }
+
+    def test_put(self, client, server_config, provide_metadata):
+        response = client.put(
+            f"{server_config.rest_uri}/datasets/metadata",
+            json={
+                "controller": "node",
+                "name": "drb",
+                "metadata": {"seen": False, "saved": True, "user": {"xyzzy": "plugh"}},
```

Comment:
I realize that "this is just a unit test", but isn't seen-false and saved-true an invalid configuration?  I recommend that we stick to valid configurations, so that we don't have to fix this test if the something starts checking the semantics.

Reply:
Technically it's not invalid, although I can't imagine why the dashboard would set them that way so it's certainly "implausible" if perhaps not quite "inconceivable".

Reply:
How could it ever be invalid?? It's client-side metadata!  It's not for us to say what is valid and what is not!....  (Still, how can the user have requested that his data be saved sight-unseen?  A hollow voice says 'PLUGH'.)